### PR TITLE
fix: user state management

### PR DESCRIPTION
**Pull request by SamSamskies**
resolves https://github.com/saving-satoshi/saving-satoshi/issues/114

I fixed the bug, but can whoever reviews this make sure the signup and login flows are correct? I'm not 100% sure if they are correct just by looking at the mockups.

I used `use-local-storage-state` to access values from localStorage in order to make sure components using localStorage values re-render whenever the values in localStorage changes.

~~A lot of the changes are just formatting that was done by Prettier. I prettified every file that I touched which is a strategy I've used on codebases that are not fully prettified. If anyone does not agree with this strategy and would prefer to review a smaller changeset without all the formatting changes, let me know and I can simplify the PR.~~

I also removed any unused variables in files I touched. I'd really recommend that we set up a linting rule to not allow unused variables.

### CODE REVIEW

File: components/Fonts.tsx
Comment (SamSamskies) [inline components/Fonts.tsx:28]: There were warnings about these invalid attributes so I removed them. I don't noticed anything that breaks by removing them, but let me know if they are supposed to be there for some reason.

**Comment by ecurrencyhodler** [comment]
Tested the PR. The site no longer crashes when I:

1. Open site in incognito window
2. Click on avatar icon on top right
3. Click I've copied and backed up my code
4. Click on avatar icon on top right again
5. Click Clear saved progress

**Comment by SamSamskies** [comment]
This bug still exists on master, so this PR will still need to be reviewed once I rebase. Not sure if I'll get to it today. Probably tomorrow.

**Comment by SamSamskies** [comment]
@ecurrencyhodler you should probably functionally test this again. The rebase was gnarly.

@afterburn or whoever else please review whenever you get a chance. Thanks.

**Comment by afterburn** [comment]
> @afterburn we might want to rethink using default exports. I don't really care about it too much, but I tend to agree with the points made here https://basarat.gitbook.io/typescript/main-1/defaultisbad. On the last couple of large codebases I worked on my team switched to using a convention of never using default exports and it was nice.

Sure, I'm not strongly opinionated on this topic, so if you and the rest of the team feel this improves our codebase and DX then I think we should do it.

File: package.json
Comment (SamSamskies) [inline package.json:19]: I removed the TS compile check from the pre-commit hook. It wasn't working correctly. Spent a little time to try and fix it, but couldn't figure it out and decided to punt on it.

**Comment by SamSamskies** [comment]
oops I broke the build. right after I took out the typecheck from the pre-commit hook 😂. at least the Vercel deployment check caught it.

**Comment by SamSamskies** [comment]
k ready for review now

**Comment by afterburn** [comment]
Maybe its just the deployment that's broken, but when I create a new user, then refresh the page, it gets me into a state I cannot get out of unless I clear my localStorage manually (see video). All seems to work just fine when I check out this branch locally though.

https://user-images.githubusercontent.com/10757768/216434405-ce6e6a9f-65a9-4bd6-88b8-d5816b783195.mp4

Can someone else help verify? Maybe @GBKS or @ecurrencyhodler? If it turns out that just the deployment is broken then LGTM.

**Review by afterburn — APPROVED** [review]

**Comment by SamSamskies** [comment]
Wth. I wonder why does that only happen on the deployed at? 🤔

**Comment by SamSamskies** [comment]
I was able to reproduce locally by running the app in prod mode instead of dev mode. Don't merge until I fix the bug.

**Comment by SamSamskies** [comment]
Should be fixed now. Please test all user flows thoroughly before merging. 🙏

**Comment by SamSamskies** [comment]
So prod mode statically builds some of the components which was causing the bug. So far not liking the Next.js app directory magic at all.

**Review by afterburn — APPROVED** [review]

**Comment by ecurrencyhodler** [comment]
This is the flow of testing. I'm not sure this is accurate or expected:

1. I go to website and click on empty profile on top right.
2. I'm initially prompted to save my progress.
3. I copy and click "I've copied and backed up my code".
4. I exit the screen and click on the profile (which now has an avatar) on the top right again.
5. I'm prompted to save code or to log out now.
6. After I click "log out" I click on empty profile and am prompted to enter code.
7. I click "load my progress"
8. Then it says progress loaded and I'm able to continue.

If this is all expected user flow then this is a tACK.

**Comment by ecurrencyhodler** [comment]
tACK. Okay I think to me the UX flows makes sense after going through prod and comparing the differences.

The language on the profile screen is confusing though at this state. There is nothing to load. But I'll either submit a PR or Issue to fix it. 
Screenshot 2023-02-02 at 9 13 06 PM


Sam are we g2g with merge?

**Comment by GBKS** [comment]
Thanks for implementing this properly, much appreciated.

@ecurrencyhodler what's confusing about the text? Let's make sure copy change are first addressed in design before doing PRs.

**Comment by SamSamskies** [comment]
@ecurrencyhodler ship it 🚀

**Comment by ecurrencyhodler** [comment]
Okay will do Christoph.

**Review by ecurrencyhodler — APPROVED** [review]

**Comment by ecurrencyhodler** [comment]
@SamSamskies looks like there's one conflict that needs to be resolved before getting merged.

**Comment by SamSamskies** [comment]
@ecurrencyhodler rebased. I'd recommend functionally testing again before merging.

**Comment by ecurrencyhodler** [comment]
I'll definitely test. But we should have another review before merge too since commits were made after the last approval.

**Comment by SamSamskies** [comment]
I only made one tiny commit to fix the build after the rebase https://github.com/saving-satoshi/saving-satoshi/pull/119/commits/d3c3139befc636a69a66eb33c863bc74864d5920

**Comment by ecurrencyhodler** [comment]
tACK

**Pull request merged**
